### PR TITLE
Enum type added to the types metadata table

### PR DESCRIPTION
**Pull request by LindsayWray**
This pr addresses issue #5174  and offers an alternative solution. 

@Mause suggested to output an additional field in the underlying table (duckdb_types) instead, which the pg_... view consumes to provide a postgres compatible interface and therefore compatible with SQLAlchemy. So I’ve added the enum type to duckdb_types. 

image

14 nov. additions: 
The pg_enum view now shows the metadata of enum types similar to Postgres. To get the enum data for this view, I have added an extra column to the duckdb_types table containing the possible values of each enum type.
Some small changes in the format_pg_type function and pg_type view were also required to make this functionality resemble the Postgres results as much as possible.

### CODE REVIEW

File: test/sql/keywords/keywords_in_expressions.test [hunk 1]
```diff
@@ -11,12 +11,18 @@ PRAGMA enable_verification;
 statement ok
 CREATE SCHEMA "SCHEMA";
 
-statement ok
+statement error
 CREATE TYPE "ENUM" AS ENUM('ALL');
```
Comment (Mytherin) [inline test/sql/keywords/keywords_in_expressions.test:15]: Is this change necessary?

File: test/sql/pg_catalog/sqlalchemy.test [hunk 2]
```diff
@@ -4,6 +4,9 @@
 
 # https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/dialects/postgresql/base.py
 
+statement ok
+CREATE TYPE greeting AS ENUM('hi', 'bonjour', 'konnichiwa', 'howdy')
+
```
Comment (Mytherin) [inline test/sql/pg_catalog/sqlalchemy.test:9]: There's another query on the bottom that concerns enums (under `# load enums`):

```sql
SELECT t.typname as "name",
	-- no enum defaults in 8.4 at least
	-- t.typdefault as "default",
	pg_catalog.pg_type_is_visible(t.oid) as "visible",
	n.nspname as "schema",
	e.enumlabel as "label"
FROM pg_catalog.pg_type t
		LEFT JOIN pg_catalog.pg_namespace n ON n.oid = t.typnamespace
		LEFT JOIN pg_catalog.pg_enum e ON t.oid = e.enumtypid
WHERE t.typtype = 'e'
```

Does that now also work/produce the same result as Postgres?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Some comments:

File: test/sql/keywords/keywords_in_expressions.test (hunk 1 above)
Comment (LindsayWray) [reply test/sql/keywords/keywords_in_expressions.test:15]: As a side effect it's now no longer possible to name your ENUM as ENUM. I could perhaps look into making this possible again if you like?

File: test/sql/pg_catalog/sqlalchemy.test (hunk 2 above)
Comment (LindsayWray) [reply test/sql/pg_catalog/sqlalchemy.test:9]: I will look into it

File: test/sql/keywords/keywords_in_expressions.test (hunk 1 above)
Comment (Mytherin) [reply test/sql/keywords/keywords_in_expressions.test:15]: Ah that’s alright, that makes sense - thanks for explaining

**Review by pdet — APPROVED** [review]
Hey Lindsay, many thanks for the PR, it looks great!

Out of curiosity, I see that 'ENUM' becomes an invalid name for an enum, is it only on caps, or 'enum'/'EnUm' are also now invalid?

(I don't even remember if we are case sensitive for enum types although I implemented that (: )

**Comment by LindsayWray** [comment]
> Hey Lindsay, many thanks for the PR, it looks great!
> 
> Out of curiosity, I see that 'ENUM' becomes an invalid name for an enum, is it only on caps, or 'enum'/'EnUm' are also now invalid?
> 
> (I don't even remember if we are case sensitive for enum types although I implemented that (: )

it's case insensitive, enum has been added to the catalog and is now like a reserved typename

File: test/sql/pg_catalog/sqlalchemy.test (hunk 2 above)
Comment (LindsayWray) [reply test/sql/pg_catalog/sqlalchemy.test:9]: I've updated the pr and this test now gives the same result as postgres.

**Review by Mytherin — APPROVED** [review]
Thanks for the updates! Looks great.

**Comment by cpcloud** [comment]
cc @saulpw, looks like we can properly support this in ibis now!

**Comment by Mause** [comment]
Looking forward to testing this 😁

**Comment by Mause** [comment]
Doesn't look like this is working as expected quite just yet, I'm looking into it

**Comment by Mause** [comment]
Looks like `duckdb_columns()` might not properly handle user defined types? https://github.com/duckdb/duckdb/blob/9cfe9c42d8e427cb4dd1a439256bce10f0530d74/src/function/table/system/duckdb_columns.cpp#L211 vs https://github.com/duckdb/duckdb/blob/9cfe9c42d8e427cb4dd1a439256bce10f0530d74/src/function/table/system/duckdb_types.cpp#L85-L98

**Comment by LindsayWray** [comment]
> Looks like `duckdb_columns()` might not properly handle user defined types?
> 


I suggest to make this into a separate issue, because I suspect more of these system tables will require similar additions to fully support enums.

**Comment by Mause** [comment]
> > Looks like `duckdb_columns()` might not properly handle user defined types?
> > 
> 
> 
> I suggest to make this into a separate issue, because I suspect more of these system tables will require similar additions to fully support enums. 

Sounds likely yes
